File: tests/test_dag_solver.cpp
/*
 * Copyright (c) 2022 Authors:
 *   - Trevor Phillips <trevphil3@gmail.com>
 *
 * Distributed under the CC BY-NC-SA 4.0 license
 * (See accompanying file LICENSE.md).
 */

#include <gtest/gtest.h>

#include <memory>

#include "core/logic_gate.hpp"
#include "core/solver.hpp"
#include "core/sym_hash.hpp"
#include "dag_solver/dag_solver.hpp"
#include "eval_solver.hpp"
#include "hashing/sym_md5.hpp"
#include "hashing/sym_ripemd160.hpp"
#include "hashing/sym_sha256.hpp"

namespace preimage {

TEST(DAGSolverTest, SolveSHA256) {
  std::shared_ptr<Solver> solver = std::make_shared<DAGSolver>();
  for (int d = 4; d <= 8; ++d) {
    std::shared_ptr<SymHash> hasher = std::make_shared<SHA256>(64, d);
    for (int i = 0; i < 5; ++i) {
      const bool solved = evaluateSolver(solver, hasher);
      EXPECT_TRUE(solved);
    }
  }
}

TEST(DAGSolverTest, SolveMD5) {
  std::shared_ptr<Solver> solver = std::make_shared<DAGSolver>();
  for (int d = 8; d <= 12; ++d) {
    std::shared_ptr<SymHash> hasher = std::make_shared<MD5>(64, d);
    for (int i = 0; i < 5; ++i) {
      const bool solved = evaluateSolver(solver, hasher);
      EXPECT_TRUE(solved);
    }
  }
}

TEST(DAGSolverTest, SolveRIPEMD160) {
  std::shared_ptr<Solver> solver = std::make_shared<DAGSolver>();
  for (int d = 8; d <= 12; ++d) {
    std::shared_ptr<SymHash> hasher = std::make_shared<RIPEMD160>(64, d);
    for (int i = 0; i < 5; ++i) {
      const bool solved = evaluateSolver(solver, hasher);
      EXPECT_TRUE(solved);
    }
  }
}

TEST(DAGSolverTest, NegatedBitAssignments) {
  const LogicGate g = LogicGate::fromString("A 3 1 -2");
  const SymRepresentation problem({g}, {1, -2}, {3});
  std::unordered_map<int, bool> assignments;
  assignments[3] = true;
  assignments[-2] = true;
  DAGSolver solver;
  EXPECT_THROW({ solver.solve(problem, assignments); }, std::invalid_argument);
}

TEST(DAGSolverTest, UnsatisfiableProblem) {
  const LogicGate g = LogicGate::fromString("A 3 1 -2");
  const SymRepresentation problem({g}, {1, -2}, {3});
  std::unordered_map<int, bool> assignments;
  assignments[3] = true;
  assignments[1] = true;
  assignments[2] = true;
  DAGSolver solver;
  EXPECT_THROW({ solver.solve(problem, assignments); }, std::runtime_error);
}

}  // end namespace preimage

File: tests/test_dag_solver.py
"""
Copyright (c) 2022 Authors:
    - Trevor Phillips <trevphil3@gmail.com>

Distributed under the CC BY-NC-SA 4.0 license
(See accompanying file LICENSE.md).
"""

import pytest

import cryptosym
from cryptosym import DAGSolver

HASH_AND_DIFFICULTY = (
    (cryptosym.SymMD5, 4),
    (cryptosym.SymRIPEMD160, 4),
    (cryptosym.SymSHA256, 4),
    (cryptosym.SymSHA512, 4),
)


class TestDAGSolver:
    @pytest.mark.parametrize("data", HASH_AND_DIFFICULTY)
    def test_preimage(self, data: tuple[cryptosym.SymHash, int], helpers):
        cryptosym.utils.seed(14)

        hash_class, difficulty = data
        hasher = hash_class(64, difficulty=difficulty)
        hash_output = hasher()
        hash_hex = cryptosym.utils.hexstr(raw_bytes=hash_output)
        problem = hasher.symbolic_representation()
        solver = DAGSolver()
        assert solver.solver_name() == "DAG Solver"

        solution = solver.solve(problem, hash_output=hash_output)
        preimage = helpers.build_input_bytes(solution, problem.input_indices)
        assert hash_output == hasher(preimage)

        solution = solver.solve(problem, hash_hex=hash_hex)
        preimage = helpers.build_input_bytes(solution, problem.input_indices)
        assert hash_output == hasher(preimage)

        assignments = helpers.build_assignments(hash_output, problem.output_indices)
        solution = solver.solve(problem, bit_assignments=assignments)
        preimage = helpers.build_input_bytes(solution, problem.input_indices)
        assert hash_output == hasher(preimage)

    def test_negated_bit_assignments(self):
        g = cryptosym.LogicGate("A 3 1 -2")
        problem = cryptosym.SymRepresentation([g], [1, -2], [3])
        assignments = {3: True, -2: True}
        solver = DAGSolver()
        with pytest.raises(ValueError):
            _ = solver.solve(problem, bit_assignments=assignments)

    def test_unsatisfiable_problem(self):
        g = cryptosym.LogicGate("A 3 1 -2")
        problem = cryptosym.SymRepresentation([g], [1, -2], [3])
        assignments = {3: True, 1: True, 2: True}
        solver = DAGSolver()
        with pytest.raises(RuntimeError):
            _ = solver.solve(problem, bit_assignments=assignments)
